chips/stm32f429zi/src/dma1.rs

```diff
@@ -829,6 +797,159 @@ pub trait StreamClient {
+                            },
+                        )+
+                    }
+                });
+            }
+
+            fn set_peripheral_address(&self) {
+                self.peripheral.map(|pid| {
+                    match pid {
+                        $(
+                            Dma1Peripheral::$Peripheral => unsafe {
+                                DMA1.registers.$sXpar.set($address);
+                            },
+                        )+
+                    }
+                });
+            }
+
+            fn set_peripheral_address_increment(&self) {
+                self.peripheral.map(|pid| {
+                    match pid {
+                        $(
+                            Dma1Peripheral::$Peripheral => unsafe {
+                                DMA1.registers.$sXcr.modify($SXCR::PINC::CLEAR);
+                            },
+                        )+
+                    }
+                });
+            }
+
+            fn set_memory_address(&self, buf_addr: u32) {
+                self.peripheral.map(|pid| {
+                    match pid {
+                        $(
+                            Dma1Peripheral::$Peripheral => unsafe {
+                                DMA1.registers.$sXm0ar.set(buf_addr);
+                            },
+                        )+
+                    }
+                });
+            }
+
+            fn set_memory_address_increment(&self) {
+                self.peripheral.map(|pid| {
+                    match pid {
+                        $(
+                            Dma1Peripheral::$Peripheral => unsafe {
+                                DMA1.registers.$sXcr.modify($SXCR::MINC::SET);
+                            },
+                        )+
+                    }
+                });
+            }
+
+
+            fn set_data_width_for_peripheral(&self) {
+                self.peripheral.map(|pid| match pid {
+                    $(
+                        Dma1Peripheral::$Peripheral => {
+                            self.stream_set_data_width(Msize(Size::Byte), Psize(Size::Byte));
+                        },
+                    )+
+                });
+            }
+
+            fn set_transfer_mode_for_peripheral(&self) {
+                self.peripheral.map(|pid| match pid {
+                    $(
+                        Dma1Peripheral::$Peripheral => {
+                            self.stream_set_transfer_mode(TransferMode::Fifo(FifoSize::Full));
+                        },
+                    )+
+                });
+            }
+        }
+    };
+}
+
+/// List of peripherals managed by DMA1
+dma1_peripheral! (
+    USART3_TX (DMA1_Stream3, Stream3, Channel4, s3cr, s3par, s3m0ar, S3CR, MemoryToPeripheral, usart::USART3.get_address_dr()),
+    USART3_RX (DMA1_Stream1, Stream1, Channel4, s1cr, s1par, s1m0ar, S1CR, PeripheralToMemory, usart::USART3.get_address_dr()),
+);
+
+
```

Comment:
I am wondering if there is anyway around *not* introducing `dma1_peripheral!` macro?

DMA is central to how Tock interacts with peripherals. When developing the UART code, I had to spend time stepping through between DMA code, UART code and its interaction with capsules and user-space.

As we add more peripherals (I2C, SPI, etc.,) in case there is an issue, debugging and stepping through this macro will become challenging, as we won't be able to see the expanded context in the debugger.

Reply:
Well, that's a good point.

Firstly, I tried to implement without macro, but found that this would make lots of duplicate codes since because current dma implementation depends on chip-specific peripheral configuration. I'm interested in adding other stm family board as well. That's why I decided to use macro.

One idea to debug may be `cargo expand`. By using this command, we can expand macro, and using expanded code for debugging, we may easily debug the code. But I don't have such experience, so not sure my idea will work.

Reply:
I would like to get feedback from Tock core team members on this. 

One approach that could work is to split this part of the code into modules and conditionally include the modules. That should preserve the debugging information (though there will be some duplication between the modules).

Reply:
Yes. Allowing duplication might be a better idea for the first step, and then, thinking about reducing duplication with less macro.

Anyway, I'd like to get their feedback too. This is my first contribution, so I'd like to follow their policy as much as possible.